packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+    const isNoUncheckedIndexedAccess = compilerOptions.noUncheckedIndexedAccess;
+
+    /**
+     * Checks and reports a destructuring node and and its matching type annotation.
+     *
+     * _Ignores non-destructuring nodes._
+     */
+    function checkParam(param: TSESTree.Node, paramType: TSESTree.Node): void {
+      if (
+        param.type === AST_NODE_TYPES.ObjectPattern &&
+        paramType.type === AST_NODE_TYPES.TSTypeLiteral
+      ) {
+        return checkObject(param, paramType);
+      }
+
+      if (
+        param.type === AST_NODE_TYPES.ArrayPattern &&
+        paramType.type === AST_NODE_TYPES.TSTupleType
+      ) {
+        return checkTuple(param, paramType);
+      }
+
+      // do nothing otherwise
+    }
+
+    function checkObject(
+      node: TSESTree.ObjectPattern,
+      typeAnnotation: TSESTree.TSTypeLiteral,
+    ): void {
+      const remainingProperties: RemainingProperties = new Map();
+      const dynamicProperties: DynamicProperties = new Set();
+
+      // collect used properties
+      for (const property of node.properties) {
+        // bail on a rest element
+        if (property.type === AST_NODE_TYPES.RestElement) {
+          return;
+        }
+
+        const valueType = services.getTypeAtLocation(property.value);
+
+        // bail if misses; this is a type-error
+        if (tsutils.isIntrinsicErrorType(valueType)) {
+          return;
+        }
+
+        const memberKey = getStaticMemberAccessValue(property, context);
+
+        // collect dynamic keys which we failed to statically analyzed
+        if (memberKey === undefined) {
+          dynamicProperties.add({ property });
+          continue;
+        }
+
+        remainingProperties.set(memberKey, { property });
+      }
+
+      // remove used type members, report on anything that's unused
+      for (const member of typeAnnotation.members) {
+        if (member.type === AST_NODE_TYPES.TSIndexSignature) {
+          checkIndexSignature(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+
+        if (
+          member.type === AST_NODE_TYPES.TSMethodSignature ||
+          member.type === AST_NODE_TYPES.TSPropertySignature
+        ) {
+          checkStaticMember(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+      }
+    }
+
+    function checkTuple(
+      param: TSESTree.ArrayPattern,
+      typeAnnotation: TSESTree.TSTupleType,
+    ): void {
+      for (const [index, member] of typeAnnotation.elementTypes.entries()) {
+        // bail on rest type `[...string[]]`
+        if (member.type === AST_NODE_TYPES.TSRestType) {
+          return;
+        }
+
+        const property = param.elements.at(index);
+
+        if (property === undefined) {
+          reportOnMember(member, { type: 'key', key: String(index) });
+          continue;
+        }
+
+        if (
+          // bail in case of `[, ...]`
```

Comment:
[Docs] Nit: the `...` looking like a rest/spread confused me 😅. Can a different placeholder be used? Maybe:

  ```suggestion
          // bail in case of `[,]` empty slots
```